lib/arm/arm_lifter.ml

```diff
@@ -322,20 +322,22 @@ let lift_bits mem ops (insn : bits_insn ) =
     let s = exp_of_op src in
     let i24 = int32 24 in
     let i8 = int32 8 in
-    let umask = int32 0xff0000 in
-    let lmask = int32 0xff00 in
     let rev =
       let open Bil in
       s              lsl i24 lor
       s              lsr i24 lor
-      (s land umask) lsr i8  lor
-      (s land lmask) lsl i8
+      (extract 23 16 s) lsr i8  lor
+      (extract 15 8 s) lsl i8
     in
     exec [assn (Env.of_reg dest) rev] cond
   | `REV16, [|`Reg dest; src; cond; _|] ->
     let s = exp_of_op src in
-    let i16 = int32 16 in
-    let rev = Bil.(s lsl i16 lor s lsr i16) in
+    let i8 = int32 8 in
+    let rev = Bil.(((extract 7 0 s) lsl i8) lor 
+                   ((extract 15 8 s) lsr i8) lor
+                   ((extract 23 16 s) lsl i8) lor
+                   ((extract 31 24 s) lsr i8)
+                  ) in
```

Comment:
Semanitcs notes:

The code is not well-typed and `rev` is always `0`.  The expression `extract 7 0 s` has type `i8` (i.e., an 8-bit word). Thus you're shifting 8-bit words left or right by eight bits (getting zeros always).  When you `lor` N 8-bit words, you will still get an 8-bit word. Finally, you assing the `dest` variable (which is a 32-bit register) an 8-bit word. And booms.

Instead of using shifts and ors, you can use `extract` and `concat` which lets you express the semantics _exactly_ as in the reference manual. 

Style notes: 

1) Do not leave `in` `) in`, `)`, or any other purely grammatical/service words on a single line. Each line of code, shall be a line of code. The exception is `begin/end` and other structural expressions, such as while, for, etc

2) Do not put superfluous parentheses. They clutter the code and confuse as it is not clear what were the author's intentions.  

In this case, parenthesis around extract are acceptable, but keep in mind that function applications bind tighter than infix operators. Since in this case the infix operator is fully made of ASCII characters, it is really easy to confuse it with function application especially when there is no syntax highlighting, e.g., `extract 7 0 s lsl i8` is hard to parse without highlighting, thoug it has the same `(extract 7 0 s) lsl i8` meaning.